content/blog/blog-2020-07-01-kudo-dependencies.md

```diff
@@ -0,0 +1,95 @@
+
+<!-- more -->
+
+## KudoOperator Task
+
+KUDO operators already have a mechanism to deal with installation dependencies called [plans, phases, and steps](plans.md) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies. As of KUDO 0.15.x, KUDO supports a `KudoOperator` task which allows you to specify a dependency on other operators. At the very least it requires the name of the operator to be installed. You can also provide an operator version (defaults to the latest one).
+
+Here a simple example of a task specifying a dependency on the community Zookeeper operator version 0.3.0 (which will install [zookeeper 3.4.14](https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/README.md))
+
+```yaml
+tasks:
+  - name: deploy-zookeeper
+    kind: KudoOperator
+    spec:
+      package: zookeeper
+      operatorVersion: 0.3.0
+```
+
+This task can be used as part of e.g. `deploy` plan same as any other:
+
+```yaml
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: prereqs
+        strategy: parallel
+        steps:
+          - name: first
+            tasks:
+              - deploy-zookeeper
+      - name: main
+        strategy: parallel
+        steps:
+          - name: second
+            tasks:
+              - deploy-main
+```
+
+As with other tasks, KUDO will make sure this task is healthy before moving to the next one. _"Healthy"_, in the case of the operator means that its `deploy` plan has finished successfully. Note, that any transitive dependencies of the zookeeper operator itself will also be resolved and installed in the right order. `KudoOperator` closely mimics the `kubectl kudo install` command semantics allowing to additionally specify the `appVersion` (defaults to most recent one) and the `instanceName` (will be generated by KUDO by default).
+
+## Dependency Parametrization
+
+Overall we wanted to encourage operator composition by providing a way of operator encapsulation. In other words, operator users should not be allowed to arbitrarily modify the parameters of embedded operator instances. The higher-level operator should define all parameters that its direct dependency operators need. If a child operator needs to be parametrised a parameter file can be specified using the `parameterFile` field. Let's take a look at an example. Suppose, we have two operators (_parent_ and _child_) and _child_ has a required parameter `PASSWORD` which is empty by default. First, the parent operator needs to specify a  parameter file for the child and reference it in the corresponding `KudoOperator` task:
```

Comment:
wanted -> want

Can we use some other parameter than `PASSWORD`? Maybe `USERNAME`? I'd like to get rid of the idea that passwords should be stored in Params...